File: react-front/src/services/supabaseAuthService.js
import { supabase } from '../config/supabase';

// ========================================
// SERVICIO DE AUTENTICACIÓN CON SUPABASE
// ========================================

export const authService = {
  // ========================================
  // REGISTRO E INICIO DE SESIÓN
  // ========================================

  // Registrar usuario con email y contraseña
  async signUp(email, password, metadata = {}) {
    const { data, error } = await supabase.auth.signUp({
      email,
      password,
      options: {
        data: metadata, // nombre, apellido, etc.
      },
    });

    if (error) throw error;
    return data;
  },

  // Iniciar sesión con email y contraseña
  async signIn(email, password) {
    const { data, error } = await supabase.auth.signInWithPassword({
      email,
      password,
    });

    if (error) throw error;
    return data;
  },

  // Iniciar sesión con Google
  async signInWithGoogle() {
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: 'google',
      options: {
        redirectTo: window.location.origin + '/callback',
      },
    });

    if (error) throw error;
    return data;
  },

  // Iniciar sesión con GitHub
  async signInWithGitHub() {
    const { data, error } = await supabase.auth.signInWithOAuth({
      provider: 'github',
      options: {
        redirectTo: window.location.origin + '/callback',
      },
    });

    if (error) throw error;
    return data;
  },

  // Cerrar sesión
  async signOut() {
    const { error } = await supabase.auth.signOut();
    if (error) throw error;
  },

  // ========================================
  // GESTIÓN DE SESIÓN
  // ========================================

  // Obtener usuario actual
  async getCurrentUser() {
    const { data: { user }, error } = await supabase.auth.getUser();
    if (error) throw error;
    return user;
  },

  // Obtener sesión actual
  async getSession() {
    const { data: { session }, error } = await supabase.auth.getSession();
    if (error) throw error;
    return session;
  },

  // Verificar si hay usuario autenticado
  async isAuthenticated() {
    const session = await this.getSession();
    return session !== null;
  },

  // ========================================
  // RECUPERACIÓN DE CONTRASEÑA
  // ========================================

  // Enviar email para resetear contraseña
  async resetPassword(email) {
    const { data, error } = await supabase.auth.resetPasswordForEmail(email, {
      redirectTo: window.location.origin + '/reset-password',
    });

    if (error) throw error;
    return data;
  },

  // Actualizar contraseña
  async updatePassword(newPassword) {
    const { data, error } = await supabase.auth.updateUser({
      password: newPassword,
    });

    if (error) throw error;
    return data;
  },

  // ========================================
  // ACTUALIZACIÓN DE PERFIL
  // ========================================

  // Actualizar datos del usuario
  async updateProfile(updates) {
    const { data, error } = await supabase.auth.updateUser({
      data: updates,
    });

    if (error) throw error;
    return data;
  },

  // ========================================
  // LISTENERS DE AUTENTICACIÓN
  // ========================================

  // Suscribirse a cambios de autenticación
  onAuthStateChange(callback) {
    return supabase.auth.onAuthStateChange((event, session) => {
      callback(event, session);
    });
  },
};

export default authService;

File: CheckAuto2/app/src/main/java/com/tecsup/checkauto/service/SupabaseAuthService.kt
package com.tecsup.checkauto.service

import com.tecsup.checkauto.config.SupabaseConfig
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.auth.providers.Google
import io.github.jan.supabase.auth.user.UserInfo
import io.github.jan.supabase.auth.user.UserSession
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Servicio de autenticación con Supabase
 */
object SupabaseAuthService {
    private val client = SupabaseConfig.client
    
    // ========================================
    // REGISTRO E INICIO DE SESIÓN
    // ========================================
    
    /**
     * Registrar usuario con email y contraseña
     */
    suspend fun signUp(
        email: String, 
        password: String,
        nombre: String? = null
    ): UserInfo? {
        // Nota: redirectTo se configura en Supabase Dashboard
        // La URL de redirección se establece en Authentication → URL Configuration
        // La app móvil interceptará automáticamente la URL de la web mediante App Links
        client.auth.signUpWith(Email) {
            this.email = email
            this.password = password
            if (nombre != null) {
                this.data = buildJsonObject {
                    put("nombre", nombre)
                }
            }
        }
        return client.auth.currentUserOrNull()
    }
    
    /**
     * Iniciar sesión con email y contraseña
     */
    suspend fun signIn(email: String, password: String): UserInfo? {
        client.auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
        return client.auth.currentUserOrNull()
    }
    
    /**
     * Cerrar sesión
     */
    suspend fun signOut() {
        client.auth.signOut()
    }
    
    // ========================================
    // GESTIÓN DE SESIÓN
    // ========================================
    
    /**
     * Obtener usuario actual
     */
    fun getCurrentUser(): UserInfo? {
        return client.auth.currentUserOrNull()
    }
    
    /**
     * Obtener sesión actual
     */
    fun getCurrentSession(): UserSession? {
        return client.auth.currentSessionOrNull()
    }
    
    /**
     * Verificar si hay usuario autenticado
     */
    fun isAuthenticated(): Boolean {
        return client.auth.currentUserOrNull() != null
    }
    
    /**
     * Obtener ID del usuario actual
     */
    fun getCurrentUserId(): String? {
        return client.auth.currentUserOrNull()?.id
    }
    
    /**
     * Obtener email del usuario actual
     */
    fun getCurrentUserEmail(): String? {
        return client.auth.currentUserOrNull()?.email
    }
    
    // ========================================
    // RECUPERACIÓN DE CONTRASEÑA
    // ========================================
    
    /**
     * Enviar email para resetear contraseña
     */
    suspend fun resetPassword(email: String) {
        client.auth.resetPasswordForEmail(email)
    }
    
    /**
     * Actualizar contraseña
     */
    suspend fun updatePassword(newPassword: String) {
        client.auth.updateUser {
            password = newPassword
        }
    }
    
    // ========================================
    // ACTUALIZACIÓN DE PERFIL
    // ========================================
    
    /**
     * Actualizar datos del usuario
     */
    suspend fun updateProfile(nombre: String? = null) {
        client.auth.updateUser {
            if (nombre != null) {
                data = buildJsonObject {
                    put("nombre", nombre)
                }
            }
        }
    }
    
    // ========================================
    // LISTENERS
    // ========================================
    
    /**
     * Observar cambios en el estado de autenticación
     */
    fun sessionStatus() = client.auth.sessionStatus
}

